django_celery_beat/models.py

```diff
@@ -87,13 +87,7 @@ def from_schedule(cls, schedule):
         spec = {'event': schedule.event,
                 'latitude': schedule.lat,
                 'longitude': schedule.lon}
-        try:
-            return cls.objects.get(**spec)
-        except cls.DoesNotExist:
-            return cls(**spec)
-        except MultipleObjectsReturned:
-            cls.objects.filter(**spec).delete()
-            return cls(**spec)
+        return cls.objects.get_or_create(**spec)[0]
```

Comment:
This can still raise MultipleObjectsReturned, why is not handling that better?

Reply:
There is a uniqueness constraint covering all fields that go into `spec` so the database should not be able to contain more than a single match at any time.

Reply:
I see your point, but even the previous logic seems terrible tbh.  

As an example, lets say I have 2 different schedules for 2 different tasks, one running every sunrise and one running every sunset (just using solarschedule as the example but this applies to any of them).  Now, lets say that my boss comes over and says all tasks running at sunset should be changed to sunrise for only the next week for whatever business reason.  Instead of modifying a single schedule, I now have to go through and modify potentially hundreds of different periodic tasks to change their schedule... and then I have to remember which ones I changed so I can change them back later.  

It seems like we should be allowing duplicate schedules.  I'm really curious why this limitation ever existed in the first place?

Reply:
I'm digging into where all from_schedule is used to see if we can sort this out some, feel free to do the same.

Reply:
> Now, lets say that my boss comes over and says all tasks running at sunset should be changed to sunrise for only the next week for whatever business reason. Instead of modifying a single schedule, I now have to go through and modify potentially hundreds of different periodic tasks to change their schedule... and then I have to remember which ones I changed so I can change them back later.

I was thinking of the same scenario earlier. I think this feature also has downsides: Once you have multiple schedules with the same values, you'll have a hard time telling them apart when creating new jobs.

Personally, I see the value in a unique constraint but also the pain with not supporting to change a schedule into something that already exists.

To fully support the scenario that you have in mind, it would take another layer of indirection, I think.

Reply:
> I'm digging into where all from_schedule is used to see if we can sort this out some, feel free to do the same.

There is a single place outside the test suite:
https://github.com/celery/django-celery-beat/blob/db68a4997420e141b002b555d0d58e98f32a6a4f/django_celery_beat/schedulers.py#L171

Reply:
Yea, I'm reading through the celery scheduler code those inherit from to work out the logic on what scenarios this call stack actually happens.

Reply:
So, based on what I've looked through, the whole reason for all this is that when you use the django settings to set up the schedules, it deletes/inserts those into the DB.  However, instead of having a flag for "inserted from django conf" or whatever, they compare it to what's already in the DB and treat the schedule items as if they are unique constraints.  Still kind of playing around with this, but if that's the case then the deletes may be necessary for now, although there are probably several scenarios where things can get borked.  Still playing around with this though.